control-userlayer/src/main/java/org/oskari/control/userlayer/CreateUserLayerHandler.java

```diff
@@ -1,222 +1,251 @@
-                log.warn("Rename failed in temp directory", newFile.getName(), " ->  ", newName);
-            } else {
-                // update path
-                file.setSavedTo(renameLocation.getPath());
-                log.debug("File renamed ", newFile.getName(), " ->  ", newName);
+    private List<File> unZip(FileItem zipFile) throws ActionException {
+        try (InputStream in = zipFile.getInputStream();
+                ZipInputStream zis = new ZipInputStream(in)) {
+            List<File> files = new ArrayList<>();
+            int filesInZip = 0;
+            // Name all the files we find from zip with the same pattern
+            // baseName.{ext} where {ext} is taken from the original filename
+            // Use a 24 letter random UUID as baseName
+            String baseName = UUID.randomUUID().toString().substring(0, 24);
+            ZipEntry ze;
+            while ((ze = zis.getNextEntry()) != null) {
+                filesInZip++;
+                if (filesInZip > MAX_FILES_IN_ZIP) {
+                    // safeguard against infinite loop, userlayers shouldn't have this many files in any case
+                    break;
+                }
+                if (ze.isDirectory()) {
+                    continue;
+                }
+                String ext = getFileExt(ze.getName());
+                if (ext == null) {
+                    continue;
+                }
+                String name = baseName + '.' + ext;
+                File file = writeToTempFile(zis, name);
+                files.add(file);
             }
+            return files;
+        } catch (IOException e) {
+            throw new ActionException("Failed to unzip file", e);
         }
-        return file;
     }
 
-    @Override
-    public void handleAction(ActionParameters params) throws ActionException {
-
-        // stop here if user isn't logged in
-        params.requireLoggedInUser();
-
-        final String target_epsg = params.getHttpParam(PARAM_EPSG_KEY, "EPSG:3067");
-        final String source_epsg = params.getHttpParam(PARAM_SOURCE_EPSG_KEY, defaultSourceEpsg);
-        try {
-
-            // Only 1st file item is handled
-            RawUpLoadItem loadItem = getZipFiles(params);
-
-            // Checks file size
-
-            File file = unZip(loadItem.getFileitem());
+    private String getFileExt(String name) {
+        int i = name.lastIndexOf('.');
+        if (i < 0 || i > name.length() - 1) {
+            return null;
+        }
+        return name.substring(i + 1);
+    }
 
-            if (file == null) {
-                log.error("Couldn't find valid import file in zip file");
-                throw new ActionParamsException("invalid_file");
-            }
+    private File writeToTempFile(InputStream in, String name) throws IOException {
+        File file = new File(getTempDir(), name);
```

Comment:
Each file will get it's own temp directory? So having files with same filename work instead of throwing an exception about existing file?

Reply:
Nope, `getTempDir()` currently always returns the same dir (systems temp dir). However having one directory for each upload would probably be better. The way this is supposed to work right now is that the basename should be unique enough (the first 24 letters of a random UUID) so that the file names won't collide within the temp directory (and they get deleted rather quickly)

Reply:
Or is createTempFile(),getParentFile() the java.io,tmp folder so everything is saved to the tmp root?

Reply:
Yes it is. Let's discuss tomorrow how we want to handle the temp files and I'll make the required changes.